### ISSUE TITLE
Alignment of multi-camera pointclouds needs work

### ISSUE THREAD

**Issue by jackjansen** [cwi-dis/cwipc#18]
We need to fix the alignment once and for all.

~~The original issue is https://gitlab.com/VRTogether_EU/cwipc/cwipc_util/-/issues/41~~ but this link is dead now.

There is an old repo for experiments at https://github.com/cwi-dis/pointclouds-alignment

New experiment data is at https://github.com/cwi-dis/cwipc_test/tree/master/pointcloud-registration-test

> Edit 20-Nov-2023: most of the text above is outdated. Kept for future reference.

Current plan of attack:

- [x] Create an algorithm that returns the approximate mis-alignment of each camera (as a distance in meters)
- [x] Implement a few of the alignment algorithms in a way that they can be run automatically
- [x] Create a registration algorithm that does something like
  -  Given the misalignment from step 1 run one or more of the alignment algorithms. 
  - Re-measure the misalignment. Check that it has gone down.
  - Repeat until happy, or no more improvements
  - record the results in `cameraconfig.json`. Also record there the current misalignment, because it is a good value for voxelization (later, during production).
- [ ] After all this is done look at the other registration ideas @Silvia024 found, and possibly also that algorithm that @nachoreimat found.
- [x] After all this is done try to automate the coarse registration with Aruco codes
- [x] After all that is done see whether we can use multiple Aruco codes to handle a large number of cameras with fields of view that don't fully overlap.
- [ ] If (big if) the registration procedure works with point clouds of people (as opposed to only working with point clouds of boxes, or not working at all) we should try applying it to the _cwipc-sxr_ captures, to see if we can get better alignment.
- [ ] In parallel to all of the steps above we should think about whether there's a paper in here somewhere, and who should be the primary person to lead this.

**Comment by jackjansen** [comment]
Current best procedure (copied from Slack):

I think I've just done the best calibration we've ever done. Keeping note here so we can try again (in other words: 
[@shirsub](https://cwidis.slack.com/team/U3G77G6CR)
 can try this procedure in vrsmall).
1. Use 4 cameras, not 3.
2. First do the coarse calibration `cwipc_calibrate --kinect --nofine` with the A4 sheet on the floor.
3. Edit cameraconfig, set correct high/low, near/far, radius, erosion=2
4. Do `cwipc_view --kinect` with nothing in view, and ensure there are no points. If there are points edit cameraconfig to fix.
5. At the centerpoint stack 3 IKEA boxes so each camera sees 2 sides (and each side is seen by two cameras)
6. `cwipc_view --kinect` and w to save a pointcloud.
7. `cwipc_calibrate --kinect --reuse --nocoarse --nograb pointcloud_xxxxxxxxx.ply --corr 0.05`
8. (The 0.05 is how far the per-camera pointclouds are apart in the grabbed image. For many years I picked lower numbers here all the time, but the trick is you should pick _higher_ numbers.)
9. Pick algorithm 6 (cumulative multiscale ICP), variant 2 (point2plane)
10. Inspect result, after - a few times to make the points smaller. If you look at the stack of boxes from above you get a very good idea of the quality.
11. The way the algorithm works at the moment means (I think) it will be happy after doing 3 steps so with the new calibration I did another grab and calibrate with `--corr 0.02` which gave me a pretty good result (at least I think so).

**Comment by jackjansen** [comment]
I think we should start looking at `opencv` to help with calibration. For example, we could use an Aruco marker (see https://docs.opencv.org/4.x/d5/dae/tutorial_aruco_detection.html) to help with detection of `(0, 0, 0)` and do the coarse calibration automatically.

This also gives us camera positions, so we can then determine the camera ordering for the next step (which pairs of cameras we should align).

We could then capture a 3D object (maybe the boxes, but possibly a person is good enough). I have the feeling that we could use the histogram of point-to-point distances to determine how well the coarse calibration worked: I'm expecting a plateau on the left of the histogram, then a drop, then a very long tail. Somewhere along the drop would seem to be a good value for the `--corr` parameter.

We then use cumulative multi scale ICP or something, and after that recompute the histogram. `corr` should now be lower.

Then I think we repeat with the other camera pairs. 

Question is what to do after all pairs have been done, because possibly the last calibration has broken the first one.

Another option could be to first compute the `corr` for each pair, and use that to determine the order of doing the aligning, but I'm not sure whether "best first" or "worst first" is the best option. I think "best first" but I'm not sure.

But I do think that after we've finished calibration we should repeat the calculation of `corr`, because this will give us the optimal cell size for voxelization.

@fonskuijk I would really like your opinion on this idea...

**Comment by jackjansen** [comment]
Okay, let's use this issue to keep note of what we've done (and found out) so far.

- There are 6 pointclouds in `cwipc_test` that have been grabbed in _vrbig_ and _vrsmall_. One of the stack of boxes, one of Jack facing the front camera and one of Jack facing between the front camera and the next camera.
- From visual inspection the registration in _vrbig_ looks pretty good, the registration in _vrsmall_ has one camera off by a few centimeters.
- There's a script `sandbox/voxelize_curve.py` (now badly named) that vowelises a point cloud by ever decreasing voxelsize. It creates a CSV file with an entry per voxelsize, showing how many points there are, how many points with 1 contributing camera, how many points with 2 contributing cameras, etc.
- Each voxelized point cloud is also saved.
- The script has been run on the two _boxes_ pointclouds, _after colorising them per camera_, i.e. after running `--filter 'colorize(1,"camera")'` on it.
- The resulting CSV files can be imported into `voxelize_curve_spreadsheet.numbers` to graph them.
- (note that various fixes were made to `cwipc_view` and `cwipc_downsample()` to make this all work somewhat)

The results are difficult to interpret. I was expecting to see some sort of a threshold effect between `cellsize > epsilon` and `cellsize < epsilon`, where `epsilon` is how good the current registration is (in other words: when the cellsize is _so_ large that points from camera 1 and camera 2 would be merged I would expect a super-linear drop in the point count.

This doesn't happen. except for very small and very large cell sizes I'm seeing a more-or-less linear decrease in the number of points (of approximate a factor of 2, as is to be expected as I'm using `sort(0.5)` as the factor between cell sizes.

<img width="602" alt="point-count-factors-per-cellsize" src="https://github.com/cwi-dis/cwipc/assets/4665100/a2fce574-4236-4dad-bc20-b82207136776">

Next I decided to look at the number of contributing cameras for each result point (after voxelizing) at each cellsize. Again, I was expecting to see something interesting at the boundary of `cellsize == epsilon`: something like a sudden increase in the number of points with two contributing cameras (at the expense of the number of points with one contributing camera).

But again no such luck: things look vaguely linear. Moreover, already at small cellsizes there are quite a few points that already have _three_ contributing cameras. This should not be possible with the boxes, I think: I think the alignment of the cameras and the boxes was such that each side could really only be seen by two cameras.

<img width="602" alt="n-camera-contributions-cumulative" src="https://github.com/cwi-dis/cwipc/assets/4665100/fe0ba220-230b-43e5-bf1c-47950313da66">

Either I am making a stupid mistake in my thinking, or there is another bug in the way `cwipc_downsample()` works. I will investigate the latter first.

   `

**Comment by jackjansen** [comment]
That was a bit of a disappointment, so far. Or actually maybe it just showed our erronous thinking:-)

Implemented the generation of registration-test pointclouds, with a settable offset for camera 2 and 4. (1 and 3 will be correct).

Created pointclouds with an error of `0.03` and `0.003`.

The results are in https://github.com/cwi-dis/cwipc_test/tree/master/pointcloud-registration-test
but the quick message is that the curves still look pretty linear. There is a clear "minimum voxelsize" where points from two cameras where points from multiple cameras start to be combined, but this "minimum voxelsize" is quite a bit smaller than the pre-determined error with which the point clouds were generated.

I could say something (based on the data for these pointclouds) like: _"As soon as more than 15% of the points come from two cameras our voxelsize is larger than the error"_ but this would just be data-fitting.

@Silvia024 can you have a look at the data?

**Comment by Silvia024** [comment]
maybe tomorrow we can chat better about that?

btw I just found this: https://github.com/IntelRealSense/librealsense/issues/10795 
and it seems quite relevant and it might be of interest. I will try to read it tonight

**Comment by Silvia024** [comment]
I found two interesting papers...
- https://www.sciencedirect.com/science/article/abs/pii/S0143816618308273 here I found interesting that they use a cube with chessboard patterns on four sides as calibration object so they know all the measurements (e.g., distance between the edge of chessboard pattern). They put the cube such that one camera capture mainly only one side of the cube. What it's interesting to me is that to verify the reconstructed point clouds, they check if two adjacent side of the cube are orthogonal; 

- https://dl.acm.org/doi/abs/10.1145/3570361.3592530?casa_token=sNaOGgeRC5IAAAAA:4_WVQOtqLz9sVcqmvRoBOMKKBsBPZX6vtJjFrNd553VoJes5l49chIb8ytQejYnJUJWjCtdPMcKW 
this one is very recent (just published at MobiCom!) and I think they are doing what we also would like to have in our system :"We propose Dynamic Camera Calibration (DCC) in MetaStream to address this limitation. The DCC module dynamically obtains and manages the camera pose in real-time and calibrates cameras without any pre-defined makers".

**Comment by jackjansen** [comment]
Updated two previous comments (the two wild ideas): they turned out not to work. I have also experimented with changing the "next cellsize factor" from `sqrt(0.5)` to `sqrt(sqrt(0.5))` but there is no appreciable bump around the known epsilon with which the point clouds are created.

I now think that the whole idea of using voxelization at successively smaller sizes and looking for a discontinuity simply doesn't work.

> To come up with a theory: voxelization is essentially a stochastic procedure, because it depends on the origin of the grid and the individual original points have no relationship to that grid. So whether or not two points are combined if they are less that `cellsize` apart is a chance: if `cellsize > 2*distance` they are definitely combined, if `cellsize < distance/sqrt(3)` they are definitely not, but the area in between is grey. @Silvia024 does this sound reasonable?

**Comment by jackjansen** [comment]
My suggestion for the next try is to compute `foreach point in cam1cloud` the minimum distance to `any point in cam2cloud`. First thing to look at is the histogram of these distances. (April 9 comment). I hope we would see a very large peak just below `epsilon` with a very long tail (all the cam1 points that have no corresponding cam2 point).

I think the KD-Tree is the data structure we want. And it appears to be supported in NumPy, example code is in https://medium.com/@OttoYu/exploring-kd-tree-in-point-cloud-c9c767095923

**Comment by jackjansen** [comment]
The idea of using KDTree and plotting a histogram of the mimic distances looks very promising.

There's a Jupyter notebook in cwipc_test, I'm going to convert it to a script.

**Comment by Silvia024** [comment]
> Updated two previous comments (the two wild ideas): they turned out not to work. I have also experimented with changing the "next cellsize factor" from `sqrt(0.5)` to `sqrt(sqrt(0.5))` but there is no appreciable bump around the known epsilon with which the point clouds are created.
> 
> I now think that the whole idea of using voxelization at successively smaller sizes and looking for a discontinuity simply doesn't work.
> 
> > To come up with a theory: voxelization is essentially a stochastic procedure, because it depends on the origin of the grid and the individual original points have no relationship to that grid. So whether or not two points are combined if they are less that `cellsize` apart is a chance: if `cellsize > 2*distance` they are definitely combined, if `cellsize < distance/sqrt(3)` they are definitely not, but the area in between is grey. @Silvia024 does this sound reasonable?
> 
> @Silvia024 @troeggla please up-vote this comment if you agree that voxelization is a dead end. Then I can remove all the generated data and spreadsheets over in `cwipc_test/pointcloud-registration-test` (which is quite a lot).

the theory seems correct to me :)
agree that maybe it's not a good strategy

**Comment by jackjansen** [comment]
I'm getting beautiful graphs, plotting (per camera pair) the cumulative distance-to-other-cloud.

![cumdist](https://github.com/cwi-dis/cwipc/assets/4665100/8e52604b-4daa-429e-a3bd-ef661e0f29ec)

**Comment by jackjansen** [comment]
All the graphs are now in cwipc_test. They look mostly usable. I'm going to change the Y axis from being counts to being fractions, that should make them even easier to interpret (I hope).

**Comment by jackjansen** [comment]
For reference in further discussion below: here is the current state of the graph for the boxes in vrsmall (with one camera about 5cm off):

![boxes_cumdist](https://github.com/cwi-dis/cwipc/assets/4665100/04dd030f-cf19-487b-bbc1-a84e114413ce)

And here is one of the Jack point clouds (the best one, the sideways one):

![jack-sideways_cumdist](https://github.com/cwi-dis/cwipc/assets/4665100/af767fd3-8c20-4787-8496-73e9332fe12e)

**Comment by jackjansen** [comment]
First thing I would like to do is convert each of these lines into a number `correspondence`. This number should be an upper bound for the registration error between the two cameras in that pair, for the given capture. If we are pretty sure the cameras have no significant overlap in field of view (pairs `(0, 3)` and `(1, 2)` in the graphs above) we return an invalid value (`NaN` or `None` or something like that).

Once we have this number we can do two things with it:
- Use it as the `--corr` parameter to the various registration algorithms we have
- After applying an algorithm we can recompute this `correspondence` number and ascertain it is lower than the previous value.

**Comment by jackjansen** [comment]
Idea: we should look at the derivative of these lines (which is actually the distance histogram data) and the second derivative. Possibly after smoothing the lines a little.

The second derivative should start positive, then go negative, then go positive again. Maybe that last point is the number we're interested in.

**Comment by jackjansen** [comment]
Unrelated idea (@Silvia024 comments please): we are currently looking at the symmetric difference between camera pairs.

Would it be better to do a per-camera comparison to all other cameras combined? I.e. not compute and graph the distance per camera pair, but in stead look at the distance of the points of camera `N` to the points of all other cameras together?

That should extend the "tail" of the histogram, probably making it more linear-ish. Also, the resulting number `correspondence` should be more useful, because it's a number that pertains to a single camera (in stead of to a pair of cameras).

**Comment by jackjansen** [comment]
> Would it be better to do a per-camera comparison to all other cameras combined? 

Implemented, and it seems to produce _much_ better results. So much better that We need to check that we are not looking at bullshit data.

Here are the graphs (for single-camera-to-all-others) that are for the same dataset as the pair graphs of yesterday:

![boxes_cumdist_one2all](https://github.com/cwi-dis/cwipc/assets/4665100/586b6107-8740-427e-8a39-9fa9200196da)

![jack-sideways_cumdist_one2all](https://github.com/cwi-dis/cwipc/assets/4665100/8f265599-3434-4a09-863a-acf96ac27a08)

![genregtest03_cumdist_one2all](https://github.com/cwi-dis/cwipc/assets/4665100/470b9251-90a8-40af-8b90-46acab891494)

**Comment by jackjansen** [comment]
And here are the histogram graphs of the same datasets:

![boxes_histogram_one2all](https://github.com/cwi-dis/cwipc/assets/4665100/6eedfd32-8d3d-4fd8-a770-f38c8f41d94b)

![jack-sideways_histogram_one2all](https://github.com/cwi-dis/cwipc/assets/4665100/8881e117-2241-4825-ad56-c6f6a2c1652a)

![genregtest03_histogram_one2all](https://github.com/cwi-dis/cwipc/assets/4665100/58ae33d6-77dd-4e54-be63-cfa91ff7af11)

**Comment by Silvia024** [comment]
> 
> The original issue is https://gitlab.com/VRTogether_EU/cwipc/cwipc_util/-/issues/41
> 
@jackjansen, I can't access to this link. It says page not found. Maybe do I not have the right to access it?

**Comment by jackjansen** [comment]
> > The original issue is https://gitlab.com/VRTogether_EU/cwipc/cwipc_util/-/issues/41
> 
> @jackjansen, I can't access to this link. It says page not found. Maybe do I not have the right to access it?

That link is dead (the gitlab site disappeared when we stopped paying for it).

**Comment by jackjansen** [comment]
I just realised that the correspondence error alone is not good enough to determine in which order we should attempt to re-align the cameras. I previously thought it would be enough: just re-align the cameras starting with the worst one.

But I now think we also want to know how many points had a distance below the correspondence error: if camera 1 has `corr=0.01, n_corr_points=20` and camera 2 has `corr=0.005, n_corr_points=20000` it's probably a much better idea to start with camera 2.

**Comment by Silvia024** [comment]
Some points from our meeting this morning.
- On our slack channel there is pinned the folder with initial papers for our literature review,
- TO DO: check this repository https://github.com/XuyangBai/awesome-point-cloud-registration and extract algorithms that could be applied to our settings;
- TO DO: list of metrics to evaluate the final registration

**Comment by jackjansen** [comment]
Yet another wild idea, this time on the "ground truth" dataset we use (both for testing and for the eventual writing of the paper).

If we start with one of the `loot` pointclouds and split it into two tiles by randomly assigning each of the points to a tile we know _exactly_ what the measured correspondence should be: it is the cell size of the original `1024*1024*1024` ply-file that we created the pointcloud from (the 8i files use integer coordinates, so we know the cellsize exactly).

**Comment by jackjansen** [comment]
> "worst aligned camera" is currently determined by multiplying measured mis-alignment by the number of points in the alignment set. As you can see in the bottom graph above this is a bit too aggressive: camera 2 has a mis-alignment that is worse than camera 1 but it wasn't selected. We should try sqrt(nAlignmentPoints) or log(nAlignmentPoints) as the factor to see whether it makes a difference.

That seems to be a good idea, with `log()` working better than `sqrt()`. Here are the results for jack-sideways, first with using `nAlignmentPoints` in the weight, then with using `log(nAlignmentPoints)`. We not only did an extra step in the latter case, but also aligned different cameras at the previous steps. You can see that the latter got to a better result eventually.

But: we may be able to get even better results through completely different means. For example: if our previous algorithm didn't get anymore improvement (the end condition of the loop above) we just re-align all cameras once in no particular order. And if that resulted in any improvement we continue with our loop.

![jack-sideways_histogram_after_step_6_nPoints](https://github.com/cwi-dis/cwipc/assets/4665100/5b788c78-e8d6-44c5-a200-079a14ed22d0)

![jack-sideways_histogram_after_step_7](https://github.com/cwi-dis/cwipc/assets/4665100/94a1b0a5-ff16-42b5-a0d2-29a9852df8ae)

**Comment by jackjansen** [comment]
> But: we may be able to get even better results through completely different means. For example: if our previous algorithm didn't get anymore improvement (the end condition of the loop above) we just re-align all cameras once in no particular order. And if that resulted in any improvement we continue with our loop.

Implemented this (looping over the other cameras after the worst camera doesn't show anymore improvement), and it turns out not to be a good idea.

In essence the only thing it does is to move some "badness" from one camera to another. So, each time through the loop the camera that's being re-aligned will improve a bit, but other cameras will get a bit worse. I'm not sure I understand why this happens...

And there's a problem about determining when to terminate.

**Comment by Silvia024** [comment]
> Observations:
> 
> * All captured datasets seem to end with alignment errors of around `3.5 mm`. I wouldn't be surprised if this turns out to be the resolution of the Kinects at the given distance between camera and subject (around `2 m` I think). Anyone willing to compute this number?

I think so, I found that it should be: "~ 3 mm per pixel x by y at 2 m"

> * A result of the previous point is that I don't think the ICP variants that take color into account are going to provide any benefit for us. @Silvia024 @ashutosh3308 what do the scientists think?
need to check this. I don't have an answer now

> * It think `ICP-point2plane` may be able to do better than the `3 mm` we are getting now, because the "plane" will be determined using some form of interpolation, and therefore the camera resolution at the given distance should play less of a role. @Silvia024 @ashutosh3308 what do the scientists think? And more importantly: is there anything about this in the literature?

From a quick research, I see that `ICP-point2point` has generally worse performance than `ICP-point2plane` but is fast to compute. 
Will check better in the literature

**Comment by ireneviola** [comment]
Ideas:
-first use geometry-only `ICP`, and then try with color to see if it gives any improvement

**Comment by Silvia024** [comment]
This is the path for calibration algorithms: python/cwipc/scripts/cwipc_calibrate/calibrator.py (https://github.com/cwi-dis/cwipc_util/blob/12ed6b2e2cf783a18194109f784b748535955d8a/python/cwipc/scripts/cwipc_calibrate/calibrator.py)

**Comment by jackjansen** [comment]
Implemented ICP_Point2Plane, but it seems to be performing slightly _worse_ than ICP_Point2Point.

I can see two possible reasons for this:

1. Our analysis (the correspondence computing) is not good enough (or actually: its precision is not good enough). The analysis basically uses Point2Point distance, so I can imagine that the values after ICP_Point2Point and ICP_Point2Plane both fall within the precision of the analysis.
2. The Point2Plane realignment estimates normals for each point, and this estimation is done using two magic values that determine how many neighbouring points are used to estimate the normals. The two magic values are `distance` and `nNeigbhour`. I have no idea where the values of `0.02` and `30` that we currently use come from. They may be wildly incorrect for our use case (human bodies).

I think the best way forward is to first create the loot-based test data set, described above.

**Comment by jackjansen** [comment]
Implemented part of 

> And from a practical point of view: Maybe now is the time to turn cwipc_calibrate and the sandbox/registration_error_* code into a few classes, 

and also 

> Note for the future: at the end of the re-registration run we have a correspondence. We should voxelize the pointcloud to this number (or possibly some function f() of this number). The expectation is that this vozelized pointcloud will nicely have all the "synthesised cameras" for the tile overlaps. 

The results are somewhat disappointing: for reasonable choices of `f()` (I tried `factor*correspondence`, for factors `sqrt(0.5)`, `1`, `sqrt(2)`, 2) there are not many points in the "intermediate" tiles. About `10%` or so when using `1*correspondence`. I have to go up to a fact or `4` or so before a significant number of points get into the new "synthesised intermediate camera" tiles. Even for the boxes dataset.

I have no idea why this happens, nor a good idea of how we could investigate it.

**Comment by jackjansen** [comment]
Interesting results also on fine-calibration.

Created a dataset `vrsmall-worse`after tilting three cameras in three different planes (x,y,z). The resulting point clouds are clearly off.

The fixer managed to make them all upright, but the's where the good news stops. They're still off by quite a bit, and moreover (and worse): _the analysis algorithm produces way too optimistic values_.

That is the first problem to fix. It s clearest from examining the graphs in `vrsmall-worse/boxes` and comparing the computed correspondence values to what a human can see from the plots: the values are way too optimistic.

The second thing to think about is the floor: whether we should include the floor in the capture. vrsmall-worse/boxes shows a reason to do so: it would probably have forestalled that the camera 4 point cloud (the green one) is positioned "too high". But on the other hand: having a lot of "floor points" matched may lead to the overall algorithm to be happy too sone ("I've matched up all these floor points, who cares about a few person points").

**Comment by jackjansen** [comment]
>  the analysis algorithm produces way too optimistic values.

Let's start by analysing this problem. Here's the initial distances histogram for `vrsmall-worse/boxes`:

![boxes_histogram_one2all](https://github.com/cwi-dis/cwipc/assets/4665100/4eede298-335a-492f-a3fb-240e0a3f8e74)

Visually it is clear that both camera 1 (blue) and 2 (orange) have a correspondence error of approximately `0.1`, but the analysis algorithm gives a value of around `0.04`.

Also, camera 8 (red) seems to have preciously few points in the correspondence set, making that result not very trustworthy. Similar to a lesser extent for camera 4 (green).

**Comment by jackjansen** [comment]
Wild idea: part of the problem is that we are computing correspondence on the _whole_ point clouds **S** (source, the points from the camera that we are interested in) and **D** (destination, all the other camera point clouds combined).

If we could in stead compute the subset point clouds **Soverlap** and **Doverlap** consisting of the points that could reasonably be expected to be in the overlap between S and D this should help the computation of correspondence a lot: For each point in **Soverlap** we compute the distance to the nearest point in **Doverlap**, then compute mean and stddev of those distances. Those two numbers should allow is to compute a good candidate for the correspondence (probably something like `mean + stddev/2` or so, but I'll leave that to the mathematicians:-).

Question to the scientists: does this make sense so far?

If the answer is _yes_, then here is another wild idea on how to compute **Doverlap** (and the same would work for **Soverlap**).

```
nNeighbour = 4 (or maybe 3, or 8, but some reasonably small value)
For each point in D add a field usage_count
for each point in S:
    compute the nNeighbour nearest neighbours in D
    for each of those nNeighbour neighbours
        increment usage_count
Remove all points from D that have usage_count == 0
Remove all points from D that have usage_count > nNeighbour
```

Rationale: the points in `D` with `usage_count == 0` are apparently very far away from any point in `S` (no point in `S` selected them as a close neighbour), so we can get rid of them. On the other hand, the points with `usage_count > some threshold` are apparently very _close_ to `S`, because many points in `S` that do not have any "natural neighbours" in `D` have selected them as one of the closest. So we should get rid of these points as well, because they will skew the mean and stddev later. Using `nNeighbour` as `some threshold` is a wild guess.

Scientists: does this sound reasonable?

**Comment by jackjansen** [comment]
I guess today is a wild-idea-day:-)

> Got started on playing with Aruco markers. Created a dataset vrsmall-noreg that has a capture of aruco-markers.

I was thinking on how we could create images to search from the point clouds (as opposed to using the captured images "aux data", because we don't always have these, and also because as stated above it is difficult to convert the image `(u, v)` coordinates back to spatial `(x, y, z)` coordinates).

What if we create three planes, based on the *camera* coordinate system: `x=0`, `y=0` and `z=0`. On each of these planes we make two distinct projections, for example for the `x=0` plane we make one image where we project all points with `x>=0` and ignore points with `x<0`, etc. (Actually: we need only 5 planes, because for `z=0` one of the planes would be empty as the camera can't see behind itself).

I guess we can initialise each of the planes to something funny (green?) and hope the aruco-finding algorithm will work around that.

**Comment by jackjansen** [comment]
And if it turns out that the previous idea doesn't work: we could try to find the floor, and then project just the points that are "close to" the floor onto an image that is parallel to the floor.

Finding the floor could be done using `open3d.segment_plane`. In case there are more planes in the point cloud (for example wall or ceiling) we could find a couple of planes. There is sample code for iteratively finding planes in https://github.com/yuecideng/Multiple_Planes_Detection

**Comment by jackjansen** [comment]
Turning the pointclouds into images turns out to be complicated. Part of the problem is the different coordinate systems of open3d, opencv and cwipc (left-handed, right-handed, etc). Part of the problem is that we're using map-color-to-depth, so there is not a lot of resolution left in the pointcloud (basically the D-camera resolution, which is `640x576` is the `vrsmall-noreg` capture. This isn't good enough to detect the aruco markers, at least not when projecting to the "camera plane".

The aruco-detector had no problem with the RGB images (`1280-720`) so maybe we should capture the coarse calibration pointclouds with map-depth-to-color?

We know that we're looking at planes anyway, when we're looking for the markers.

And, actually, this may be true also for the fine calibration: having more intermediate points on the "surface" of the subject may be a benefit that outweighs the negative effect of having "invented" points (as mentioned above in 

> I just realised that we should always use map color to depth for our use case

That point still holds for production, I think, but maybe not for registration.

**Comment by jackjansen** [comment]
Paused working on the aruco markers for now. 

Implemented a manual coarse calibration algorithm (basically the same as what we have: select the four coloured corners of the calibration target).

This works, the resulting pointcloud and the correspondence graph are in `pointcloud-registration-test/vrsmall-noreg/capture`.

Here is the graph:

![pointcloud-0001](https://github.com/cwi-dis/cwipc/assets/4665100/a7f48c8e-1347-4050-b96f-d06ea1d927e3)

The correspondence errors are too optimistic, but visual inspection of the graph shows that something like `0.05` is probably the correct number.

But: these pointclouds are _so_ big (500Kpoint per camera) that the analysis is very expensive: it takes about a minute on my M1 Mac, and I didn't wait for it on the Intel Mac. That's a shame, because for our common use case (capturing in a room where the walls are close enough to be captured by the cameras) this would actually be a very useful test capture...

**Comment by jackjansen** [comment]
In case we decide to go for detecting the Aruco markers in the RGB image and then finding the corresponding points in the point cloud: here are some interesting links about how the conversion of image pixel coordinates to 3D points can be done using librealsense API's:

https://github.com/IntelRealSense/librealsense/issues/8221
https://medium.com/@yasuhirachiba/converting-2d-image-coordinates-to-3d-coordinates-using-ros-intel-realsense-d435-kinect-88621e8e733a
https://github.com/IntelRealSense/librealsense/issues/11031

**Comment by jackjansen** [comment]
Back to fine registration.

There is a thinking error in the idea that _only_ looking at the correspondence between each `(camera-N, all-cameras-except-N)` is good enough.

This is shown clearly in a new dataset `offline-boxes`. See there for the full details, but here is a screenshot of the calibration it managed to come up with:

<img width="862" alt="Screenshot 2024-01-01 at 23 55 34" src="https://github.com/cwi-dis/cwipc/assets/4665100/8bd6eafe-195c-411f-97f0-d97bb3876a45">

Visually, it is clear what the **problem 1**  is: the green and blue camera _together_ (or the red and brown camera together) need to be moved to make things better. But looking at each camera in isolation is never going to work: if we would move the green camera to better align with the red one we would completely lose the alignment with the blue camera.

There is also a **problem 2** with the analysis. Here is what our code thinks of the alignment it came up with:

![captured-boxes-3_histogram_after_step_6](https://github.com/cwi-dis/cwipc/assets/4665100/c59cac8e-d883-4c85-8c0c-60c8fe7e4dee)

Those numbers (`14 mm` to `18 mm`) are far too optimistic.

For **problem 1** a potential solution is to not only look at correspondences between `(camera-N, all-cameras-except-N)` but also `(camera-N-and-M, all-cameras-except-N-and-M)` for every combination `(N, M)` where those two cameras are "closest to each other". If and `N-and-M` camera combination would end up as candidate-to-fix we would try fixing that "camera" and apply the transformation to both cameras.

But this feels like a hacker-jack solution: it seems like it might fix this problem but I'm not really sure it's really a solution and not a quick hack.

I have the feeling that tackling **problem 2** first may be better.

Ideas, anyone?

**Comment by jackjansen** [comment]
Some progress with coarse calibration.

I'm manually positioning the pointcloud in the open3d viewer so that the virtual camera is approximately where the physical camera was. I then grab the RGB and D images.

I can now detect the aruco markers in the RGB image.

But this is where the good news stops: the next step is converting the `u, v` coordinates that the aruco detector returns (coordinates in the RGB image) back to `x, y, z` coordinates in the point cloud.

I think I'm doing all the right things: I'm taking the depth from `D[v, u]` and using the camera intrinsics to de-project, and then I'm transforming with the extrinsic matrix.

But I always end up with weird origins. I've tried transposing the extrinsic matrix, I've tried mirroring `x`, or mirroring `y` and `z` (which was suggested somewhere) but nothing seems to work.

**Comment by jackjansen** [comment]
See https://github.com/isl-org/Open3D/issues/6508 for a decent description of the open3d coordinate system (RH y-UP) and its idiosyncrasies with respect to lookat and forward.

**Comment by jackjansen** [comment]
The problem may be that I "invert" the affine transform (the `extrinsic` matrix) by transposing it. That would work for a normal 3x3 matrix in our case (because we know all the matrices are size and shape preserving) but of course it doesn't work for an affine transform, I have to use another vector as the fourth column (and clear the fourth row).

Here is an explanation: http://negativeprobability.blogspot.com/2011/11/affine-transformations-and-their.html

**Comment by jackjansen** [comment]
And I found https://stackoverflow.com/questions/2624422/efficient-4x4-matrix-inverse-affine-transform which says a similar thing. Also, looking at the source code for open3d `PointCloud.create_from_depth_image` it seems to be that they're indeed doing this, but they have the eigen3d affine transform which has an invert. I guess I'll have to create that by hand.

**Comment by jackjansen** [comment]
Automatic coarse calibration based on Aruco markers is working. It's actually working so well that I have merged back into `master`, at a1b2855, so this can now be considered production-ready.

**Comment by jackjansen** [comment]
Multi-marker coarse alignment, for camera positions where not all cameras can see the `(0, 0, 0)` origin aruco marker so we use auxiliary markers, is also working good enough that I've merged it into `master`, at b923190

**Comment by jackjansen** [comment]
So, back to fine calibration or actually first to the analysis of the current calibration.

I'm working with the `offline-boxes` capture, because that shows the problem most clearly. I've created plots for all analysers that we have (`one2all`, `one2all-filtered`, `one2all-reverse`, `one2all-reverse-filtered` and `pairwise`). All the graphs are in `cwipc_test`.

The most informative graph _for this dataset_ (but note the italics) is the pairwise graph:

![captured-boxes-3_histogram_paired](https://github.com/cwi-dis/cwipc/assets/4665100/58d83ee7-9142-4a7d-97de-e88a69858d23)

The "camera numbers" here are the `or` of the two contributing cameras. As a human we can easily see that camera 1 is opposite camera 4 and camera 2 is opposite camera 8. And I also know this is correct, because I know that the cameras are placed in the order 1-2-4-8 clockwise. We can probably detect this algorithmically if we want.

As a human we can estimate the correspondences of the pairs:

- 1 to 2 (red): about 2cm
- 2 to 4 (olive): about 4mm
- 4 to 8 (turquoise): at least 4cm
- 8 to 1 (dark blue): less than 4mm

We can also see that the correspondence errors that the current "algorithm" (but really "quick hack" is a better term) has come up with are wildly wrong. Not surprising: the current "algorithm" works by finding the peak in the histogram and then move right until we get below a value that is less than `0.5*peak`.

I will experiment with `mean` and `stddev` to see if I can get some more decent numbers. Then, if they work for this dataset, they should also be tried for the captured Jack dataset.

**Comment by jackjansen** [comment]
Mean and stddev by themselves are not going to work. Here are the results for the graph above:

```
camera 3: mean=0.02291982490598397, std=0.024645493357004753, peak=0.002320735082210554, corr=0.007816463821634046
camera 5: mean=0.04401491186517467, std=0.028213745113280272, peak=0.012704797435346417, corr=0.058097526853423245
camera 9: mean=0.016755202242697633, std=0.026204388020447566, peak=0.0018051266662951993, corr=0.002804512231245586
camera 6: mean=0.015378887548555181, std=0.023343767740899458, peak=0.001824420385722404, corr=0.003573071088354493
camera 10: mean=0.048489777693837444, std=0.028316377093057312, peak=0.0021639993720285154, corr=0.08227788490063927
camera 12: mean=0.0341007288961789, std=0.023926849570313418, peak=0.01910221049639875, corr=0.04007440525818792
```

The mean for the "good pairs" (6 and 9) is far too high. 

And that is pretty logical, when you think about it: the long tails have an inordinate effect on the mean.

Next thing to try: first compute mean and stddev. then throw away all distances that are larger than (wild guess) `mean+stddev`, or maybe `2*mean`. Then compute mean and stddev on the points that remain.

The idea is that for the "bad pairs" this will throw away less of the points, but for the "good pairs" it will throw away more points.

**Comment by jackjansen** [comment]
Tried that. Also tried running the filtering multiple times, to see how the mean and stddev behave. Used the bracketing filter `[mean-stddev, mean+stddev]` on two premises:

- It sort-of feels more mathematically correct,
- it just so happens that for the "good pairs" `std > mean` so we don't throw away any "good points", while for the "bad pairs" `std < mean` so we throw away points on both sides, so running the filter successively should not change `mean` too much (while for the "good pairs" it will lower mean).

Here are the results:

```
camera 3: peak=0.002320735082210554, corr=0.007816463821634046
camera 3: 0 filters: mean=0.02291982490598397, std=0.024645493357004753, nPoint=81262
camera 3: 1 filters: mean=0.012951448749072387, std=0.010969866908687875, nPoint=66992
camera 3: 2 filters: mean=0.009752436971104068, std=0.005658494330190411, nPoint=52946
camera 3: 3 filters: mean=0.009181153840633022, std=0.0032857770409335132, nPoint=32361
camera 5: peak=0.012704797435346417, corr=0.058097526853423245
camera 5: 0 filters: mean=0.04401491186517467, std=0.028213745113280272, nPoint=44935
camera 5: 1 filters: mean=0.04137415890726912, std=0.016101593626257675, nPoint=26525
camera 5: 2 filters: mean=0.04045734373520693, std=0.009445048042243229, nPoint=15533
camera 5: 3 filters: mean=0.040051235318581284, std=0.005529610745340443, nPoint=8873
camera 9: peak=0.0018051266662951993, corr=0.002804512231245586
camera 9: 0 filters: mean=0.016755202242697633, std=0.026204388020447566, nPoint=81399
camera 9: 1 filters: mean=0.006063694689290081, std=0.008691325161748644, nPoint=67977
camera 9: 2 filters: mean=0.003177805834766034, std=0.0024780598363417766, nPoint=59915
camera 9: 3 filters: mean=0.002515016672254931, std=0.0011297240188485047, nPoint=52181
camera 6: peak=0.001824420385722404, corr=0.003573071088354493
camera 6: 0 filters: mean=0.015378887548555181, std=0.023343767740899458, nPoint=93198
camera 6: 1 filters: mean=0.006850735824829795, std=0.00853339253154322, nPoint=79885
camera 6: 2 filters: mean=0.003812802111714147, std=0.0027496213518108494, nPoint=69301
camera 6: 3 filters: mean=0.003089897111975529, std=0.00140128710337592, nPoint=56900
camera 10: peak=0.0021639993720285154, corr=0.08227788490063927
camera 10: 0 filters: mean=0.048489777693837444, std=0.028316377093057312, nPoint=43545
camera 10: 1 filters: mean=0.04792050129221857, std=0.016245794338419692, nPoint=25355
camera 10: 2 filters: mean=0.04781840036709909, std=0.00938153612816915, nPoint=14709
camera 10: 3 filters: mean=0.04774644484942719, std=0.005421145457380683, nPoint=8482
camera 12: peak=0.01910221049639875, corr=0.04007440525818792
camera 12: 0 filters: mean=0.0341007288961789, std=0.023926849570313418, nPoint=73397
camera 12: 1 filters: mean=0.028625500662925147, std=0.011632248990853532, nPoint=52026
camera 12: 2 filters: mean=0.027763304899151773, std=0.0064978872262983645, nPoint=33259
camera 12: 3 filters: mean=0.027708708994364492, std=0.003789831220465292, nPoint=19513
```

This seems to be going in the right direction: the "bad pairs" (opposing cameras) have their `mean` staying put at high values. The "good pairs" have their mean going down significantly, towards what appears to be a correct value. The "not so good pairs" (3 and 12) also seem to end up at decent values.

**Comment by jackjansen** [comment]
Partial success. That is to say: this works pretty well for camera-pair measurements on the boxes:

![captured-boxes-3_histogram_paired](https://github.com/cwi-dis/cwipc/assets/4665100/c559b991-3078-4747-907e-dc24b0c4405e)

These are pretty beleivabe numbers!

Unfortunately it does *not* work well at all for the one-to-all-others measurements:

![captured-boxes-3_histogram_one2all](https://github.com/cwi-dis/cwipc/assets/4665100/dbdc5c8d-b85d-49cc-ae4e-6eef42978c61)

I think the problem is that this algorithm throws away any points that it can't match (which, in case of this dataset, includes the mismatched "edges that are sticking out".

Let's first check how the pair-wise measurements work on the other datasets.

**Comment by jackjansen** [comment]
That didn't work very well. I've now made the pair-wise measurement symmetric but this needs work: at the moment it is far too expensive.

**Comment by jackjansen** [comment]
For future reference: when we get back to finding the "best" algorithm to align the pointclouds we should look at point-to-plane ICP with a robust kernel. From https://www.open3d.org/docs/latest/tutorial/pipelines/robust_kernels.html#Vanilla-ICP-vs-Robust-ICP I get the impression that the robust kernel is a way to deal with noise. The referenced page uses generated noise, but of course our sensors are also noisy...

**Comment by jackjansen** [comment]
Copied from Slack:

Folks, in your research of registration algorithms, have you come across any that allow "pinning" of one of the variables? I.e. ask the algorithm to find the optimal transformation but specifying, for example, that the y-translation must be zero?
Because if that exists then we could do fine calibration in two steps:

- First do a fine calibration of the empty capture. This will align all the floors. Then assure that the floors also fall in the plane y=0.
- Next do the fine calibration with boxes or people or whatever, but pin y=0.

Actually, thinking a bit more, we not only want to pin the y-translation to 0 but also the x-rotation and z-rotation. So the only free variables should be y-rotation, x-translation and z-translation.

**Comment by ireneviola** [comment]
We might want to change the loss function as to only consider a 2D error - which would effectively mean that in every iteration, the algorithm would be forced to change only the parameters that are considered in the loss function, because the others would have no impact in the error. There might be other ways of writingit as an optimization problem, though; we should check it out

**Comment by jackjansen** [comment]
> The fixer managed to make them all upright, but the's where the good news stops. They're still off by quite a bit, and moreover (and worse): _the analysis algorithm produces way too optimistic values_.

Inspecting this issue again after half a year of inactivity, but a _lot_ of actually using the current registration setup in production. The comment quoted above (from 05-Jan-2024) seems to be the main thing that is bothering us most at the moment: Often, when running `cwipc_register --fine`, the script will report that it has managed to align all point clouds to within a few millimeters. But actual inspection of the captured point cloud clearly shows that some areas (and often important areas like the head) are off by 5-10cm.

The "solution" we are currently using is to simply try again with the subject human in a different pose, and hoping for the best.

Fixing this, or at least showing the operator something (for example a graph of the p2p distance distribution) from which they can tell this has happened is at the moment of paramount importance.

**Comment by jackjansen** [comment]
Once we have addressed the issue above (the bad numbers coming out of our analysis) my feeling is that we should move to a "mixed" upper strategy. Right now our upper strategy is either pairwise or one-to-all-others, but maybe we should first do one round of pairwise, and after that a round of one-to-all-others.

If we do the pairwise round in the right order (i.e. most overlapping pair first) I think that should get us out of the "local minimum problem" with the boxes.

The right order should be easy to compute: for each pair, compute the upper bound of the percentage/fraction of points that could possibly overlap. High-to-low is the right order.

**Comment by jackjansen** [comment]
Fixed those issues by modifying cameraconfig. The new one is attached here: 

[cameraconfig.json](https://github.com/user-attachments/files/18945426/cameraconfig.json)

With this cameraconfig we took a capture just after Thong clapped his hands (slightly after `ts=1732710676049`.


And here is the histogram plot of the final resulting distances between each camera and all others:

![Image](https://github.com/user-attachments/assets/dd78abe4-f88c-4d25-9da5-6703b63f7e06)

We can see a number of (completely different) things from this plot:

- The histogram plot is not the right one to show, we want cumulative because it would make it much easier (we hope) to interpret the results.
- The correspondence error computed for camera 4 is far too optimistic.

We need to fix the first bullet. Then we need to decide what we do first:
- Fix the second bullet, or
- Have a "human inspection" outer algorithm, where the user decides which camera to try fixing next.

**Comment by jackjansen** [comment]
I think the only improvement that could still be done is to do different steps (the outer algorithm):

1. Start with a synthetic floor (optionally). Call this the "current set"
2. Find the camera that is nearest to the current set.
3. Align it to the current set.
4. add it to the current set.
5. Repeat until there are no cameras left.

Maybe after this we could do one more pass over all the cameras but I'm not sure this is worthwhile.

**Comment by jackjansen** [comment]
It turns out that that previous comment is indeed needed, for some situations. Because we still sometimes have the issue that the cameras are "pairwise aligned" (the problem we saw last year with the boxes datasets).

**Comment by jackjansen** [comment]
Hmm, thinking out loud: maybe we should use a much smaller correspondence to the alignment step, basically reducing the points taken into account far too much, but then at least we know that the only points taken into account are points that are probably correct...

**Comment by jackjansen** [comment]
It may be that our analyzer functions are still too optimistic. I'm seeing a case where the distribution plots of the distance clearly doesn't correspond to reality.

**Comment by jackjansen** [comment]
Indeed! And it's the floor that is making it so optimistic! If I remove the floor I am getting much more realistic numbers.

**Comment by jackjansen** [comment]
That led to an insight:

- The `correspondence` number we compute in our analyzer is based on the _minimum_ possible distance. We do all sorts of tricks with repeated filtering and adding stddev and what not, but it is still based on "the first peak in the distribution of point distances", which is the minimum realistic distance. So this number is in essence _minimum-realistic-correspondence_
- When we include the floor, this will give us a whole lot of points that will be close to other floor points, and they are immune to rotation around the Y axis or translation along the X or Z axis.
- So including the floor is a bad idea with the current analysis algorithms.

But really: removing the floor is a hack. And, moreover, if we don't include the floor in the capture the alignment algorithm can't use it to "stabilise" the orientation along the Y axis of the point cloud.

And removing the floor only for the computation of the `correspondence` might work, but it is still a hack: for example, it wouldn't work for the "flat sides" of the boxes that are also giving use problems.

What I think we _really_ need is a way to compute _maximum realistic correspondence_: any points that are further from the other point cloud than this number we can confidently ignore.

We've been looking at the distribution of the distances, and at the cumulative graph of that, which is in essence the integral of it. Maybe we should look at the derivative and see whether that gives us any interesting information.

cc @Silvia024

**Comment by jackjansen** [comment]
Read through a lot of the old comments. https://github.com/cwi-dis/cwipc/issues/18#issuecomment-1850900806 struck me as that we might want to look at it again.

And there are an awful lot of comments to the effect that "our analysis algorithm is too optimistic"....

**Comment by jackjansen** [comment]
Implemented plotting the derivative of the distance distribution graph. It doesn't appear to show a lot of interesting stuff.

Next step: study https://www.open3d.org/html/tutorial/t_pipelines/t_icp_registration.html because there are a few interesting points in there:

- it suggests that Point2Plane requires normals, so us using it at this point (without having computed normals) may be stupid.
- maybe we should switch to multi scale ICP, so we can do away with the computing the upper bound of the correspondence.

**Comment by jackjansen** [comment]
Experimented with the Open3D Tensor ICP algorithms, but they seem to be much slower (at least on the CPU). Punting on it for now. But keeping the code, because it is a path towards multi-scale ICP.

**Comment by jackjansen** [comment]
I now think that the floor is key, but we need to handle it better.

One idea:
- Create the synthetic floor, as we do now, by projecting all points to `Y=0`.
- Align all point clouds to that.
- Create the real floor, by throwing away the synthetic floor and filtering all points with `Y < epsilon`.
- Continue with the algorithm as it was.

Another idea, maybe to be tried after that one: one of the problems is that the floor is _so_ dense that floor points drown out other points, for RMSE and such. We could try "thinning" the floor, maybe by removing a large fraction of the floor points.

**Comment by jackjansen** [comment]
Another wild idea!

After the floor-creation above, we do a 2D step, where we use all points that are not actually on the floor, but we project them to a fixed Y value.

Then we do the final step.

**Comment by jackjansen** [comment]
First idea (2-step process to find the floor) implemented, and works. The other 2 ideas (floor-thinning and 2D) not implemented yet.

I want to first see if it helps if we find a better value for the correspondence we pass to the ICP algorithms.

Currently the algorithms return the minimum reasonable correspondence. Want to see what number we get if we subsequently remove all source points with distances less than `corr + stddev`, and then repeat the correspondence computation.

The idea is that there is a cluster of points at that low correspondence points (the floor, for instance) and then hopefully another cluster at a higher correspondence.

**Comment by jackjansen** [comment]
One of the issues specifically with the realsense cameras is that the depth data is "wavy". When we do the coarse calibration we depend on depth information of one exact pixel, the corner point of the Aruco marker (and if we don't have that one pixel we will search for the closest pixel that we have). (And that for all 4 corners).

This may lead to significant "tilting" of the Aruco marker plane, and to mis-alignment, and to problems later on.

We should experiment with a different way to find the depth value of the Aruco corners. Maybe average over a `3x3` or `5x5` depth pixel grid?

Also, an intermediate step (between coarse and fine) `--floor` could be used to flatten the floor. If this is done with the floor-aligner using ICP-with-color and nobody in the frame (but with the Aruco marker) this might set the stage nicely for the fine alignment.

**Comment by jackjansen** [comment]
There is something very fishy going on with the coarse registration for realsense cameras. The Aruco marker is `17.5 x 17.5 cm`. But here is the `--verbose` output of one of the measurements:

```
cwipc_register: camera 0: show RGB image with detected marker
cwipc_register: camera 0: _find_markers: Auruco-IDs: [0], 2D-Arease: [[[348.0, 439.0], [312.0, 444.0], [312.0, 385.0], [349.0, 385.0]]]
cwipc_register: camera 0: find_markers: marker 0, corner 0: u,v,d=(348, 439, 2038)
cwipc_register: camera 0: find_markers: marker 0, corner 0: 3d-point=(-0.9528622031211853, 0.21710489690303802, 2.0369999408721924)
cwipc_register: camera 0: find_markers: marker 0, corner 1: u,v,d=(312, 444, 1862)
cwipc_register: camera 0: find_markers: marker 0, corner 1: 3d-point=(-0.9773123264312744, 0.21341519057750702, 1.8609999418258667)
cwipc_register: camera 0: find_markers: marker 0, corner 2: u,v,d=(312, 385, 1850)
cwipc_register: camera 0: find_markers: marker 0, corner 2: 3d-point=(-0.9731330275535583, 0.04060521349310875, 1.8539999723434448)
cwipc_register: camera 0: find_markers: marker 0, corner 3: u,v,d=(349, 385, 2026)
cwipc_register: camera 0: find_markers: marker 0, corner 3: 3d-point=(-0.9440015554428101, 0.04417876899242401, 2.0260000228881836)
cwipc_register: find_markers_all_tiles: camera 0: marker ids: dict_keys([0])
cwipc_register: find_markers_all_tiles: camera 0: marker 0:  3D corner: (-0.9528622031211853, 0.21710489690303802, 2.0369999408721924), distance to next: 0.1777285067932375
cwipc_register: find_markers_all_tiles: camera 0: marker 0:  3D corner: (-0.9773123264312744, 0.21341519057750702, 1.8609999418258667), distance to next: 0.17300218002029447
cwipc_register: find_markers_all_tiles: camera 0: marker 0:  3D corner: (-0.9731330275535583, 0.04060521349310875, 1.8539999723434448), distance to next: 0.1744861895785414
cwipc_register: find_markers_all_tiles: camera 0: marker 0:  3D corner: (-0.9440015554428101, 0.04417876899242401, 2.0260000228881836), distance to next: 0.17350203165119785
cwipc_register: camera 1: show RGB image with detected marker
cwipc_register: camera 1: _find_markers: Auruco-IDs: [0], 2D-Arease: [[[308.0, 299.0], [328.0, 347.0], [296.0, 378.0], [273.0, 327.0]]]
cwipc_register: camera 1: find_markers: marker 0, corner 0: u,v,d=(308, 299, 1911)
cwipc_register: camera 1: find_markers: marker 0, corner 0: 3d-point=(-1.0297971963882446, -0.20476090908050537, 1.9170000553131104)
cwipc_register: camera 1: find_markers: marker 0, corner 1: u,v,d=(328, 347, 2031)
cwipc_register: camera 1: find_markers: marker 0, corner 1: 3d-point=(-1.0271828174591064, -0.06264939159154892, 2.0339999198913574)
cwipc_register: camera 1: find_markers: marker 0, corner 2: u,v,d=(296, 378, 1870)
cwipc_register: camera 1: find_markers: marker 0, corner 2: 3d-point=(-1.0334938764572144, 0.03383205458521843, 1.8580000400543213)
cwipc_register: camera 1: find_markers: marker 0, corner 3: u,v,d=(273, 327, 1793)
cwipc_register: camera 1: find_markers: marker 0, corner 3: 3d-point=(-1.0641131401062012, -0.11227403581142426, 1.7940000295639038)
cwipc_register: find_markers_all_tiles: camera 1: marker ids: dict_keys([0])
cwipc_register: find_markers_all_tiles: camera 1: marker 0:  3D corner: (-1.0297971963882446, -0.20476090908050537, 1.9170000553131104), distance to next: 0.18409640597123902
cwipc_register: find_markers_all_tiles: camera 1: marker 0:  3D corner: (-1.0271828174591064, -0.06264939159154892, 2.0339999198913574), distance to next: 0.20080950332264957
cwipc_register: find_markers_all_tiles: camera 1: marker 0:  3D corner: (-1.0334938764572144, 0.03383205458521843, 1.8580000400543213), distance to next: 0.16242084318267663
cwipc_register: find_markers_all_tiles: camera 1: marker 0:  3D corner: (-1.0641131401062012, -0.11227403581142426, 1.7940000295639038), distance to next: 0.15767184927770242
cwipc_register: camera 2: show RGB image with detected marker
cwipc_register: camera 2: _find_markers: Auruco-IDs: [0], 2D-Arease: [[[219.0, 255.0], [263.0, 278.0], [253.0, 339.0], [207.0, 322.0]]]
cwipc_register: camera 2: find_markers: marker 0, corner 0: u,v,d=(219, 255, 1591)
cwipc_register: camera 2: find_markers: marker 0, corner 0: 3d-point=(-1.0574027299880981, -0.2668946087360382, 1.5889999866485596)
cwipc_register: camera 2: find_markers: marker 0, corner 1: u,v,d=(263, 278, 1809)
cwipc_register: camera 2: find_markers: marker 0, corner 1: 3d-point=(-1.075393795967102, -0.23745621740818024, 1.8070000410079956)
cwipc_register: camera 2: find_markers: marker 0, corner 2: u,v,d=(253, 339, 1758)
cwipc_register: camera 2: find_markers: marker 0, corner 2: 3d-point=(-1.0771406888961792, -0.06162986159324646, 1.7630000114440918)
cwipc_register: camera 2: find_markers: marker 0, corner 3: u,v,d=(207, 322, 1559)
cwipc_register: camera 2: find_markers: marker 0, corner 3: 3d-point=(-1.0663880109786987, -0.09644405543804169, 1.5579999685287476)
cwipc_register: find_markers_all_tiles: camera 2: marker ids: dict_keys([0])
cwipc_register: find_markers_all_tiles: camera 2: marker 0:  3D corner: (-1.0574027299880981, -0.2668946087360382, 1.5889999866485596), distance to next: 0.22071320993486104
cwipc_register: find_markers_all_tiles: camera 2: marker 0:  3D corner: (-1.075393795967102, -0.23745621740818024, 1.8070000410079956), distance to next: 0.1812566181845225
cwipc_register: find_markers_all_tiles: camera 2: marker 0:  3D corner: (-1.0771406888961792, -0.06162986159324646, 1.7630000114440918), distance to next: 0.20821302978140688
cwipc_register: find_markers_all_tiles: camera 2: marker 0:  3D corner: (-1.0663880109786987, -0.09644405543804169, 1.5579999685287476), distance to next: 0.17347947289953897
cwipc_register: camera 3: show RGB image with detected marker
cwipc_register: camera 3: _find_markers: Auruco-IDs: [0], 2D-Arease: [[[270.0, 441.0], [262.0, 385.0], [297.0, 370.0], [304.0, 421.0]]]
cwipc_register: camera 3: find_markers: marker 0, corner 0: u,v,d=(270, 441, 1895)
cwipc_register: camera 3: find_markers: marker 0, corner 0: 3d-point=(-1.1137868165969849, 0.24895404279232025, 1.8880000114440918)
cwipc_register: camera 3: find_markers: marker 0, corner 1: u,v,d=(262, 385, 1853)
cwipc_register: camera 3: find_markers: marker 0, corner 1: 3d-point=(-1.1134402751922607, 0.08012675493955612, 1.8480000495910645)
cwipc_register: camera 3: find_markers: marker 0, corner 2: u,v,d=(297, 370, 2034)
cwipc_register: camera 3: find_markers: marker 0, corner 2: 3d-point=(-1.1076630353927612, 0.039723072201013565, 2.0269999504089355)
cwipc_register: camera 3: find_markers: marker 0, corner 3: u,v,d=(304, 421, 2038)
cwipc_register: camera 3: find_markers: marker 0, corner 3: 3d-point=(-1.097619652748108, 0.20514899492263794, 2.049999952316284)
cwipc_register: find_markers_all_tiles: camera 3: marker ids: dict_keys([0])
cwipc_register: find_markers_all_tiles: camera 3: marker 0:  3D corner: (-1.1137868165969849, 0.24895404279232025, 1.8880000114440918), distance to next: 0.17350149902208023
cwipc_register: find_markers_all_tiles: camera 3: marker 0:  3D corner: (-1.1134402751922607, 0.08012675493955612, 1.8480000495910645), distance to next: 0.18359411366202733
cwipc_register: find_markers_all_tiles: camera 3: marker 0:  3D corner: (-1.1076630353927612, 0.039723072201013565, 2.0269999504089355), distance to next: 0.1673188738038414
cwipc_register: find_markers_all_tiles: camera 3: marker 0:  3D corner: (-1.097619652748108, 0.20514899492263794, 2.049999952316284), distance to next: 0.16859489982907377
```

If you look at the `distance to next:` values (which are the distances between adjacent 3D corners of the marker) they are pretty okay for the first camera, but after that they are off by sometimes as much as `4 cm`. 

The center points are computed by taking the average of the corners, so **this** could be the root cause of the unequal origins between the cameras....

**Comment by jackjansen** [comment]
That indeed seems to be part of the problem (and maybe a large part). For each of the corners, we find the depth values (the `u,v,d=` print statements above).

But this produces a wrong value, because we are using `u,v` values from the color image and we're applying them to the depth image to find `d` (possibly after applying some grid or so). But we can't do this for realsense, because we should be using `rs2_project_color_pixel_to_depth_pixel()`.

We then supply this `u,v,d` to `map2d3d`. This one does do the correct transformation on the `u,v`, but it can find wrong values because the `d` was wrong. Also, the depth value value at the given pixel may be missing, it which case it is `0`, in which case the whole function will return `0,0,0`.

We may have to add an extra auxiliary function, to map color uv to depth uv.

None of this was a problem when we used aligned frames (which we still do for Kinect, but which stopped working for relsense with the d455 "sudden jump" problem).

**Comment by jackjansen** [comment]
Noticed by @troeggla and @KarolinaWylezek while doing the guided registration for the Spring School demo:

- The "floor step" of the fine calibration does a slight offset. Maybe use color ICP here, so that the marker can be used to "keep it honest"?
- Ensure the erosion is mentioned.

**Comment by jackjansen** [comment]
Back to alignment after a pause. One thing I noticed (experimenting with Tampere tic-tac-toe, fine alignment on a frame with the two players standing over the board): sometimes the algorithm makes a mess of it. For example, at the second try camera 0 (mask 1) was moved a meter to the side.

We should disregard these. Maybe refuse any changes where not all cameras have improved (or stayed the same)?

Also, the move was done based on floor alignment of 25 or so points.

**Comment by jackjansen** [comment]
First step in the new epoch: refactoring the code.
The `Analysis` algorithms now always take 1 source point cloud and one target point cloud.
Same for the `Alignment` algorithms.
There's a `MultiAlignment` algorithm that uses these to align multiple cameras.
So all the weirdness about the algorithm implementation having to handle multiple tiles and their results is now gone.

**Comment by jackjansen** [comment]
Implemented first bullet, with `cwipc_analyse_registration --toself` and optional `--nth` parameter.

Created a `loot`, and used it to test this. The results are somewhat suspect: it seems all points have the same distance to their nearest neighbour. Maybe an artefact of the fact that `loot` started out voxelized?

Created a new `--filter noise(0.001)` which adds a random vector with length `<= 0.001` to each point.

Tested the `--toself` on this, and it seems that the results are better, but still not completely what I expected.

**Comment by jackjansen** [comment]
Next step: create a filter `simulatecams(N)` which will simulate that a point cloud has been captured by `N` cameras (spaced equidistantly on a circle around `x=z=0`).

Probably want some points between two cameras to be assigned to either of the cameras, will see if this is an extra parameter or just hard-code it.

**Comment by jackjansen** [comment]
I'm going to create a script to build these test point clouds in a predictable manner.

**Comment by jackjansen** [comment]
That seems to work pretty good. `cwipc_create_analysis_test` can tile a loot, then apply transformation to each tile (with a given distance and random direction in the `y=0` plane), then add noise to all points.

Created a point cloud with one camera moved by `10cm` (`loot-000-100-000-000-noise005.ply`) and running the analysis on it gives the numbers we expect.

Here is the point cloud:

<img width="752" alt="Image" src="https://github.com/user-attachments/assets/94814fc4-a7ea-415e-8e94-831b22eefdbb" />

**Comment by jackjansen** [comment]
The test data has been added to `cwipc_test/pointcloud-registration-test/analysis_tests`.

And I've used it to show one of the issues (maybe the important one?) with our registration!

Source tile 2 had been moved by `10 cm`, and nothing else done except add some noise.

I now try to re-align tile 2, with

```
cwipc_find_transform --dump --sourcetile 2 --targettile 253 --analyze --dump loot-000-100-000-000-noise005.ply loot-000-100-000-000-noise005.ply
```

This works absolutely wonderful:

```
Before alignment: correspondence: 0.06096984767855937, sigma: 0.030548764860235087, count: 64173, percentage: 42.53%
After alignment: correspondence: 0.0022941491300621187, sigma: 0.0012768146639165552, count: 90569, percentage: 60.03%
```

These numbers are pretty close to how I created the test data, and visual inspection of the result also looks almost perfect.

However, this works because the other 3 tiles are correct, and I'm all those three to align against.

If I try to align only to tile 1 (one of the neighbours of tile 2) I get the following:

```
Before alignment: correspondence: 0.12124953844470136, sigma: 0.09964618304317818, count: 106204, percentage: 70.39%
After alignment: correspondence: 0.05570398465476518, sigma: 0.05808374651034722, count: 97947, percentage: 64.92%
Transform filter needed: --filter 'transform44([[0.9748694581809313, -0.17142745624107789, -0.1422749688200309, 0.29267920008429926], [0.15844541269880089, 0.9824799855855932, -0.09812303052010266, 0.020277554092330986], [0.156603290836224, 0.07311432944680456, 0.9849516252734537, -0.21356242333453285], [0.0, 0.0, 0.0, 1.0]])```

With this graph:

![Image](https://github.com/user-attachments/assets/c0858364-d6a7-4f4c-a8cc-b435f6da6a10)

And here's the point cloud (only tile 1 and 2):

<img width="760" alt="Image" src="https://github.com/user-attachments/assets/6e75db55-a70b-4e14-8519-fd6995d9f059" />

So we can see:

- The number `0.121 ± 0.099` is "too large" to be considered as the measure for how "unregistered" the point clouds are.
- The alignment algorithm is too enthusiastic currently: it's moving the pointcloud by `50 cm` or so.

Problem may be that the alignment algorithm isn't told about the maximum distance at which to look for corresponding counts (maybe using `0.121` or even `0.220` would work?).

**Comment by jackjansen** [comment]
Implemented a different way of seeing how well point clouds overlap, using Open3d `evaluate_registration`. This gives us `fitness` as a fraction of the points that could be matched and `rmse` as an indication of how well they matched.

Unfortunately `fitness` doesn't seem to be very useful, it tends to be `1.0`.

Maybe forget about this for now.

**Comment by jackjansen** [comment]
Next plan: work on the multi-camera algorithms again. Plan of attack:

- Select the "best" tile to start with.
  - First idea: do `analyse(tile, ~tile)` and pick the one with the highest count.
- Select this as the result.
- Pick as the next tile the remaining one with the lowest correspondence to the result set.
- align.
- add to the result set.
- repeat.

**Comment by jackjansen** [comment]
Getting back to registration once again, just in time for the third birthday of this issue.
Started by reading through all comments here, and chanced upon https://github.com/cwi-dis/cwipc/issues/18#issuecomment-1912774786
It states that we should look at "robust kernels" for ICP, because these might be a way to forestall the outliers breaking our alignment.

The section on that in the open3d docs indeed shows exactly the problem we seem to be having with vanilla ICP: https://www.open3d.org/docs/latest/tutorial/pipelines/robust_kernels.html#Vanilla-ICP-vs-Robust-ICP shows the serious misalignment.

There's various robust kernel loss functions. The API documentation suggests two papers to check:

> The kernels implemented so far, and the notation has been inspired by the publication: “Analysis of Robust Functions for Registration Algorithms”, by Philippe Babin et al.

> For more information please also see: “Adaptive Robust Kernels for Non-Linear Least Squares Problems”, by Nived Chebrolu et al.

Someone needs to read these. @Silvia024 is this up your alley, by any chance?

**Comment by jackjansen** [comment]
I'm getting all enthusiastic about https://arxiv.org/abs/1810.01474 now!

Maybe it is not ruining our idea of a paper, but actually showing us how to write our paper. The difference with theirs is that we would be using our human point clouds.

@ireneviola @Silvia024 please comment.

We first take our 600K loot (new-style tiled) and create a dataset with a zillion perturbations to it (noise added `N`, camera positions changed `P`, camera orientation changed `O`).

We then run this dataset through the algorithms, producing something similar to their figure 2. This should give us (for each algorithm) the range of the per-algorithm tuning parameter `k` that works best.

If we can determine how `k` depends on the `N`, `P` and `O` used to create the dataset then by determining those N,P,O for a real camera setup we can find the ideal `k` (for each algorithm), and we can do a number of captures (considerably less than a zillion) with various camera systems and show the idea works.

**Comment by jackjansen** [comment]
Experimenting with `cwipc.scripts.cwipc_analyze_registration`.

`--toself` looks at nearest point distances within each tile. It gives very reasonable values of around `2mm`. Increasing `--nth` to get the n-th neighbour increases this as expected.

Without arguments (which computes per-tile distance to all other tiles: the same `2.5mm±1.5mm` for `60%` overlap of the points.

But `--pairwise` becomes weird:

```
Tiles used in source: [1, 2, 4, 8]
Alignment 0x1 to 0x2: correspondence: 0.053279, sigma: 0.061157, count: 151280, percentage: 60.62%
Alignment 0x1 to 0x4: correspondence: 0.215845, sigma: 0.095636, count: 124386, percentage: 49.84%
Alignment 0x1 to 0x8: correspondence: 0.042422, sigma: 0.065776, count: 170872, percentage: 68.47%
Alignment 0x2 to 0x1: correspondence: 0.080771, sigma: 0.111557, count: 101985, percentage: 67.62%
Alignment 0x2 to 0x4: correspondence: 0.064844, sigma: 0.094364, count: 100931, percentage: 66.92%
Alignment 0x2 to 0x8: correspondence: 0.147275, sigma: 0.083369, count: 87330, percentage: 57.90%
Alignment 0x4 to 0x1: correspondence: 0.229974, sigma: 0.098500, count: 122219, percentage: 52.50%
Alignment 0x4 to 0x2: correspondence: 0.034611, sigma: 0.048465, count: 154793, percentage: 66.49%
Alignment 0x4 to 0x8: correspondence: 0.047331, sigma: 0.062564, count: 145505, percentage: 62.50%
Alignment 0x8 to 0x1: correspondence: 0.052819, sigma: 0.088800, count: 107556, percentage: 71.26%
Alignment 0x8 to 0x2: correspondence: 0.139564, sigma: 0.060355, count: 77744, percentage: 51.51%
Alignment 0x8 to 0x4: correspondence: 0.107267, sigma: 0.119979, count: 89824, percentage: 59.51%
```

so there's something wrong with this algorithm. And this is the algorithm used in `cwipc_find_transform` to set the max distance.

**Comment by jackjansen** [comment]
> note this comment was from before the previous one, but sitting on my home machine without having been added.

I now understand why the numbers are weird: all points are taken into account when computing the `mean`, which is used as the correspondence. This is plain wrong.

Experimented with using the `median`, which looks better at first sight, but it is really also wrong when you think about it: it's just the point where `50%` of the points have a lower distance, and this doesn't really tell us anything. 

Also tried the `mode` but that's a completely useless value for a (near-)continuous value like distance between points.

Idea: we can use `mean` from the `--toself` distance (or maybe `mean+stddev`, or maybe that value for the `nth` closest point) as the bin size for a histogram when comparing two tiles. Then computing the mode using that

**Comment by jackjansen** [comment]
Current plan of attack:

- [ ] For every tile we first compute the _internal_ correspondence. 
  - This is a measure for the quality of the capture of that camera, and can be used as a lower bound when trying to match another tile to this one (or the other way around): there's no point in trying to get better than this number. 
  - This number (or this number divided by 2) can be used as the bin size of the histograms for `mode` calculations. Let's call it `target`.
  - This number is probably best computed using the `median` method: if we find the distance at which 50% of the points have a neighbour that is closer that seems like a decent guess for the capture quality.
- [ ] Now, for each tile-pair, we compute a lower bound and an upper bound for the current correspondence:
  - the lower bound is the `mode` 
  - the upper bound is the _begin of ever-continuing downward trend_ as per https://github.com/cwi-dis/cwipc/issues/18#issuecomment-3118519671

I now think that we can compute _begin of ever-continuing downward trend_ as follows (we may want to bounce it off Rob):

- step `1`:
  - Look at the derivative of the histogram computed with the bin size `target` from above.
  - If the derivative has only positive numbers(`>=0`) followed by only negative numbers (`<=0`) we are done, and the `median` is the number we want.
  - Otherwise, for step `N` we compute the histogram with a bin size of `N*target` and we try again.
  - I _think_ we can be sure that this algorithm terminates, because at some point we will have two bins at which point the first one must have a positive derivative and the second one must have a negative derivative.

**Comment by jackjansen** [comment]
On computing the internal correspondence. Let's first see if this idea makes sense.

I've look at the two real-live captures (`man...` and `woman...` and computed these numbers.

Man:
```
Alignment 0x1 self, nth=1: correspondence: 0.002721, sigma: 0.000000, count: 29250, percentage: 50.00%
Alignment 0x2 self, nth=1: correspondence: 0.002935, sigma: 0.000000, count: 33552, percentage: 50.00%
Alignment 0x4 self, nth=1: correspondence: 0.002730, sigma: 0.000000, count: 38964, percentage: 50.00%
Alignment 0x8 self, nth=1: correspondence: 0.002735, sigma: 0.000000, count: 27033, percentage: 50.00%
```

Woman:
```
Alignment 0x1 self, nth=1: correspondence: 0.002517, sigma: 0.000000, count: 31444, percentage: 50.00%
Alignment 0x2 self, nth=1: correspondence: 0.003376, sigma: 0.000000, count: 24016, percentage: 50.00%
Alignment 0x4 self, nth=1: correspondence: 0.003106, sigma: 0.000000, count: 18831, percentage: 50.00%
Alignment 0x8 self, nth=1: correspondence: 0.002811, sigma: 0.000000, count: 35615, percentage: 50.00%
```

So, based on these numbers, the man should have been captured by three cameras that were at approximately the same distance from the subject but camera 2 should have been a little further away.

And for the woman if we order the cameras by distance from the subject this order should be `1, 4, 3, 2`.

Let's look at the `cameraconfig.json` files used to capture these point clouds.

For the man, indeed camera 2 is the furthest away from the origin. And the man isn't standing exactly on the origin, but close enough.

The woman is standing almost exactly on the origin. Yes! the camera distances are exactly in that order!

**Comment by jackjansen** [comment]
Note to self: once the have the algorithms from https://github.com/cwi-dis/cwipc/issues/18#issuecomment-3124725336 we should try them on the old captures that were causing problems with alignment (such as the boxes that were pairwise-aligned, but the two pairs were misaligned). I'm expecting to see that problem show up in the _begin of downward trend_ number.

**Comment by jackjansen** [comment]
Another note: we may want to look at _KDE_, Kernel Density Estimation, as an alternative to histograms. Should ask Rob about this.

**Comment by jackjansen** [comment]
Played with the captured man and woman a bit, but inconclusive because I can't run the Kinect grabber on my Mac.

Then played with one of the Tampere captures (realsense, which I can run) and it's not making gross errors, but it isn't exactly great either.

We either need to try the interactive approach or we should go for the kernel-based ICP.
